src/jdk.unsupported/share/classes/sun/misc/Unsafe.java

```diff
@@ -718,14 +706,27 @@ public Object staticFieldBase(Field f) {
         if (f == null) {
             throw new NullPointerException();
         }
+        assertNotTrusted(f);
+        return theInternalUnsafe.staticFieldBase(f);
+    }
+
+    private static void assertNotTrusted(Field f) {
         Class<?> declaringClass = f.getDeclaringClass();
         if (declaringClass.isHidden()) {
             throw new UnsupportedOperationException("can't get base address on a hidden class: " + f);
         }
         if (declaringClass.isRecord()) {
             throw new UnsupportedOperationException("can't get base address on a record class: " + f);
         }
-        return theInternalUnsafe.staticFieldBase(f);
+        Class<?> fieldType = f.getType();
+        Class<?>[] interfaces = fieldType.getInterfaces();
+        if (interfaces != null) {
+            for (Class<?> inter : interfaces) {
+                if (inter.getName().equals("jdk.internal.lang.stable.TrustedFieldType")) {
+                    throw new UnsupportedOperationException("can't get field offset for a field of type " + fieldType.getName() + ": " + f);
+                }
+            }
+        }
```

Comment:
Given that the [`Class​::forName​(String, boolean, ClassLoader)`] method doesn’t care about whether the requested class is actually exported to the caller, it’s possible to do the following:
```suggestion
        final class Holder {
            static final Class<?> TRUSTED_FIELD_TYPE;
            static {
                PrivilegedAction<ClassLoader> getPlatformClassLoader = ClassLoader::getPlatformClassLoader;
                @SuppressWarnings("removal")
                ClassLoader platformClassLoader = AccessController.doPrivileged(getPlatformClassLoader);
                
                try {
                    TRUSTED_FIELD_TYPE = Class.forName("jdk.internal.lang.stable.TrustedFieldType",
                        false, platformClassLoader);
                } catch (ClassNotFoundException e) {
                    throw new AssertionError(e);
                }
            }
        }

        Class<?> declaringClass = f.getDeclaringClass();
        if (declaringClass.isHidden()) {
            throw new UnsupportedOperationException("can't get base address on a hidden class: " + f);
        }
        if (declaringClass.isRecord()) {
            throw new UnsupportedOperationException("can't get base address on a record class: " + f);
        }
        Class<?> fieldType = f.getType();
        if (Holder.TRUSTED_FIELD_TYPE.isAssignableFrom(fieldType)) {
            throw new UnsupportedOperationException("can't get field offset for a field of type " + fieldType.getName() + ": " + f);
        }
```

[`Class​::forName​(String, boolean, ClassLoader)`]: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/Class.html#forName(java.lang.String,boolean,java.lang.ClassLoader)